scipy/signal/spectral.py

```diff
@@ -267,121 +267,617 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
-                else:
-                    Pxx *= k/(k+1.0)
-                    Pxx[..., 0] += xft[..., 0]**2 / (k+1)
-                    Pxx[..., 1:] += (xft[..., 1::2]**2 + xft[..., 2::2]**2) \
-                                  / (k+1.0)
-
-        Pxx[..., 1:-1] *= 2*scale
-        Pxx[..., (0,-1)] *= scale
-        f = np.arange(Pxx.shape[-1]) * (fs/nfft)
+    if isinstance(window, string_types) or type(window) is tuple:
+        win = get_window(window, nperseg)
     else:
-        for k, ind in enumerate(indices):
-            x_dt = detrend_func(x[..., ind:ind+nperseg])
-            xft = fftpack.fft(x_dt*win, nfft)
-            if k == 0:
-                Pxx = (xft * xft.conj()).real
-            else:
-                Pxx *= k/(k+1.0)
-                Pxx += (xft * xft.conj()).real / (k+1.0)
-        Pxx *= scale
-        f = fftpack.fftfreq(nfft, 1.0/fs)
+        win = np.asarray(window)
+        if len(win.shape) != 1:
+            raise ValueError('window must be 1-D')
+        if win.shape[0] > x.shape[-1]:
+            raise ValueError('window is longer than x.')
+
+    if np.result_type(win,np.complex64) != outdtype:
+        win = win.astype(outdtype)
+
+    if mode == 'psd':
+        if scaling == 'density':
+            scale = 1.0 / (fs * (win*win).sum())
+        elif scaling == 'spectrum':
+            scale = 1.0 / win.sum()**2
+        else:
+            raise ValueError('Unknown scaling: %r' % scaling)
+    else:
+        scale = 1
+
+    if return_onesided is True:
+        if np.iscomplexobj(x):
+            sides = 'twosided'
+        else:
+            sides = 'onesided'
+            if not same_data:
+                if np.iscomplexobj(y):
+                    sides = 'twosided'
+    else:
+        sides = 'twosided'
+
+    if sides == 'twosided':
+        numFreqs = nfft
+    elif sides == 'onesided':
+        if nperseg % 2:
+            numFreqs = (nfft + 1)//2
+        else:
+            numFreqs = nfft//2 + 1
+
+    # Stride, detrend, apply windows
+    result = _stride_windows(x, nperseg, noverlap, axis=-1)
+    result = detrend_func(result)
+    result = _apply_window(result, win, axis=-1)
+
+    # Zero pad here, now that windowing is done
+    padShape = list(result.shape)
+    padShape[-1] = nfft - padShape[-1]
+    result = np.concatenate((result, np.zeros(padShape)), axis=-1)
+
+    # Perform the fft, don't keep redundant info
+    result = fftpack.fft(result, n=nfft, axis=-1)[...,:numFreqs]
+    freqs = fftpack.fftfreq(nfft, 1/fs)[:numFreqs]
+
+    if not same_data:
+        # All the same operations on the y data
+        resultY = _stride_windows(y, nfft, noverlap, axis=-1)
+        resultY = detrend_func(resultY)
+        resultY = _apply_window(resultY, win, axis=-1)
+        padShape = list(resultY.shape)
+        padShape[-1] = nfft - padShape[-1]
+        resultY = np.concatenate((resultY, np.zeros(padShape)), axis=-1)
+        resultY = fftpack.fft(resultY, n=nfft, axis=-1)[...,:numFreqs]
```

Comment:
and can you use PEP8 compliant names? So ``pad_shape``, ``result_y``, ``num_freqs``, etc.